include/boost/geometry/algorithms/detail/envelope/range.hpp

```diff
@@ -35,50 +48,117 @@ namespace detail { namespace envelope
 {
 
 
-struct default_expand_policy
+// implementation for simple ranges
+struct envelope_range
 {
-    template <typename Box, typename Geometry>
-    static inline void apply(Box& mbr, Geometry const& geometry)
+    template <typename Iterator, typename Box>
+    static inline void apply(Iterator first, Iterator last, Box& mbr)
+    {
+        typedef typename std::iterator_traits<Iterator>::value_type value_type;
+
+        // initialize MBR
+        initialize<Box, 0, dimension<Box>::value>::apply(mbr);
+
+        Iterator it = first;
+        if (it != last)
+        {
+            // initialize box with first element in range
+            dispatch::envelope<value_type>::apply(*it, mbr);
+
+            // consider now the remaining elements in the range (if any)
+            for (++it; it != last; ++it)
+            {
+                dispatch::expand<Box, value_type>::apply(mbr, *it);
+            }
+        }
+    }
+
+    template <typename Range, typename Box>
+    static inline void apply(Range const& range, Box& mbr)
     {
-        geometry::expand(mbr, geometry);
+        return apply(boost::begin(range), boost::end(range), mbr);
     }
 };
 
 
-template <typename ExpandPolicy, typename Iterator, typename Box>
-inline void envelope_iterators(Iterator first, Iterator last, Box& mbr)
+// implementation for multi-ranges
+template <typename EnvelopePolicy>
+struct envelope_multi_range
 {
-    for (Iterator it = first; it != last; ++it)
+    template <typename MultiRange, typename Box>
+    static inline void apply(MultiRange const& multirange, Box& mbr)
     {
-        ExpandPolicy::apply(mbr, *it);
+        typedef typename boost::range_iterator
+            <
+                MultiRange const
+            >::type iterator_type;
+
+        // initialize MBR
+        initialize<Box, 0, dimension<Box>::value>::apply(mbr);
+
+        // skip through empty geometries
+        iterator_type it = boost::begin(multirange);
+        while (it != boost::end(multirange) && geometry::is_empty(*it))
+        {
+            ++it;
+        }
+
+        // if there are still elements left, apply the envelope policy
+        if (it != boost::end(multirange))
+        {
+            // compute the initial envelope
+            EnvelopePolicy::apply(*it, mbr);
+
+            // for the remaining non-empty geometries just expand the mbr
+            // using the envelope
+            for (++it; it != boost::end(multirange); ++it)
+            {
+                if (! geometry::is_empty(*it))
+                {
+                    Box helper_mbr;
+                    EnvelopePolicy::apply(*it, helper_mbr);
+
+                    dispatch::expand<Box, Box>::apply(mbr, helper_mbr);
+                }
+            }
+        }
```

Comment:
This two loops look too complex.
You can easily just walk through it with a normal for-loop, and add one boolean variable ```initialized``` to check if you should calculate the initial envelope or expand it.
You can then also move the generic initialize to the end, when it is not yet initialized.